### PR TITLE
 chore: docker build updates, automate npm publish

### PR DESCRIPTION
**Pull request by KEGustafsson**
- base image as ubuntu 20.04
- needed apt-get packages
- nodejs script installer (node 16.x)
- mDNS enabled for node
- startup.sh script to start mDNS and SK

### CODE REVIEW

**Comment by KEGustafsson** [comment]
When SK is restart from admin webpage, then also container will restart. To keep container running while e.g. signalk-server is restarting, I could add PM2 here. This would be a bit more sophisticated method to run nodejs app.

There would 2 modifications needed
`RUN npm install pm2 -g` after nodejs installation script and
change to `startup.sh` to use pm2 `/usr/bin/pm2-docker 'bin/signalk-server --securityenabled'`

With PM2 included images sizes. 
`linux/amd64 448.77 MB`
`linux/arm64 438.79 MB`

**Comment by KEGustafsson** [comment]
In original `Dockerfile`, there is `COPY empty_file tmp*/qemu-arm-stati[c] /usr/bin/`
Is this specific for building docker with used actions?
I have used `docker buildx build` in action to build docker and above part is not needed there.

**Comment by tkurki** [comment]
For reference [Dockerfile for node:16](https://github.com/nodejs/docker-node/blob/50c2806d4af0b991ed853072741a30e5d0cb777a/16/buster/Dockerfile)

**Comment by tkurki** [comment]
One obvious difference is that the official image has yarn that we don't need.

**Comment by tkurki** [comment]
Now that there's more than just a single Dockerfile: should we put everything under a single directory `docker` for example?

**Comment by tkurki** [comment]
I think we should move forward with this, but I am just saying that the way the server is built here is kinda dumb and results in images that are not exactly 1 to 1 with what is published to npm. 

What I mean is that instead of building the server and packages from scratch (and leaving devDependencies in place!) during docker build we should install from npm the ready made packages. But then we should have a github action that installs server's deps from npm (not via npm workspace), builds, tests and publishes to npm first and only then builds the docker image, either from the files built in build step or by installing the server from npm. The latter option is more like what end users do and would also work as a test of "can we install the version we just published to npm".

**Comment by KEGustafsson** [comment]
I like the idea "can we install the version we just published to npm".

This is a bit outside of docker image build, but still very much related to it.

Every bit and pieces need to be ready before docker build is triggered, but that shouldn't be a problem if all necessary packages are available from npmjs or other locations.

For docker, there are still OS related stuff, like base image, needed additional packages, extra settings and startup stuff. Anyway, Signal K server installation could follow the same path as normal user.

Would there be then any need for `ENV IS_IN_DOCKER`?

**Comment by KEGustafsson** [comment]
> Now that there's more than just a single Dockerfile: should we put everything under a single directory `docker` for example?
> 
> 

That would be a good idea.

**Comment by tkurki** [comment]
IS_IN_DOCKER disables installing new server versions in the admin ui, so yes we want it.

**Comment by KEGustafsson** [comment]
What do you think about this approach?
[Dockerfile](https://github.com/KEGustafsson/signalk-server/blob/docker-update-npm/docker/Dockerfile)
During docker image build, there would be intermediate steps to build, (test, not included) and publish node packages.
Then build folder will be removed and SK will be installed from just published packages and pushed to docker hub.
All-in-one.

**Comment by KEGustafsson** [comment]
This was/is a study that I wanted to do to check whether it is possible/feasible. Seems to be too complicated vs github actions...
Also resulted bigger size images as you said, actually surprisingly big size, more that I expected.

How to proceed forward? Should Dockerfile be simplified to build image from npmjs packages or continue current path?
For me, I think simplified version could be more suitable and usable.

**Comment by tkurki** [comment]
So with this Dockerfile the essential changes would be a smaller image and working mdns? Mdns requires host network mode, right?

Now the docker build in Github is triggered on pushing the tag for a release. Now this will not work, because in my release workflow I
- tag locally the release with `npm version`
- run `npm run release` that pushes changes, including the release tag, to Github and creates the Github Release
- run npm publish

So the new release may or may not be published in npm and just running `npm i` will install essentially a random "whatever is latest" version in npm, BUT creating a version tagged image.

Also this approach won't be able to build docker images off master.

So there are uses to building off the local files.

So what if we (you...) back up a little: install as previously, but not from official node, and we get the advantages here.

Then we can revisit installing from npm separately:
- keep the old approach for building master image
- integrate npm publish and building the docker image for that with `npm i` in a single workflow

As for the dockerfile: is there a reason for back and fort changes between root and node - why not do everything as root in one go?

**Comment by tkurki** [comment]
For the time being I'd like to keep things the way they are re: pm2 and not add it.

https://stackoverflow.com/questions/51191378/what-is-the-point-of-using-pm2-and-docker-together

**Comment by tkurki** [comment]
Probably two dockerfiles eventually.

Node user is good, same as before and non root containers are docker best practise.

Compose could map the default 3000 port so that it is usable as such.

**Comment by KEGustafsson** [comment]
Now all root actions are done first and then changed to node.
I would keep compose at host to keep things as simple as possible. If someone wants to do hardening etc.. and control, which port are open to/from container, that is doable. I could add commented code to compose, which adds this options and gives end user easy change from host to bridge network. Ok?

**Comment by KEGustafsson** [comment]
In Github action need to defined `build-arg`, e.g for `docker buildx build`
`--build-arg TAG=${{ steps.docker_meta.outputs.tags }}`

And to use it in the Dockerfile
```
ARG TAG
RUN npm i -g signalk-server@$TAG
```
Then npm intall will use same tag as rest of the build process.

**Comment by KEGustafsson** [comment]
Dockerfile, build from Github source:
```
linux/amd64, 356 MB
linux/arm/v7, 322.05 MB
linux/arm64, 346 MB
```

Dockerfile_rel, build from npmjs source:
```
linux/amd64, 305.99 MB
linux/arm/v7, 269.85 MB
linux/arm64, 293.54 MB
```
Both amd64 and arm64 images tested and basic functionality OK.

**Comment by sbender9** [comment]
Why is this publishing to npm?

**Comment by tkurki** [comment]
One thing lead to another: Docker release images should be built like a local install from npm => the release needs to be published in npm => we can automate that here also.

**Comment by sbender9** [comment]
Ahh. Got it. I see you doing that elsewhere too. Nice.

**Comment by tkurki** [comment]
Oh damn, I just realised there's preleases to think of: https://github.com/SignalK/signalk-server/blob/master/publishing.md

The main difference is in `npm publish --tag beta`. That could probably be automated with this?

And I don't get it why we have both releasing.md and publishing.md. If we get beta publishing integrated so that you just need to push a tag name ending in `-beta.XX` let's remove publishing.md and add this info in releasing.md.

**Comment by KEGustafsson** [comment]
Can beta release be named e.g. `v1.41.4-beta` in Github?

Then `v*` release can be modified to check if tag has `-beta` ending.
Nmp script would be `npm publish --tag 1.41.4-beta` and docker would have `v1.41.4-beta`.
Bit more IF-THEN to scripting...

**Comment by tkurki** [comment]
Yes.

just to be precise: the ending needs to include the version number.

**Comment by KEGustafsson** [comment]
I'll add test if tag string contains substring `*beta*` then prerelease script will be run, otherwise normal release script will be run.
No need to check anything else for additional IF-THEN.

File: .github/workflows/build-docker-release.yml [hunk 1]
```diff
@@ -88,10 +88,18 @@ jobs:
...
+        run: echo ::set-output name=tag::$(echo ${GITHUB_REF#refs/*/})    
       - run: |
           npm install --package-lock-only
           npm ci && npm cache clean --force
-          npm publish --access public
+          if [[ "${{ steps.vars.outputs.tag }}" == *beta* ]];
+            then
+              npm publish --tag beta
+            else
+              npm publish --access public
```
Comment (tkurki) [inline .github/workflows/build-docker-release.yml]: I don't think we need `--access public` here - unscoped packages (no @something) are public by default.

Comment (KEGustafsson) [reply .github/workflows/build-docker-release.yml]: `--access public` removed from all `npm publish` places.

**Pull request merged**